pp/file/file_rpc.go

```diff
@@ -0,0 +1,117 @@
+package file
+
+import (
+	"strconv"
+	"strings"
+	"time"
+	"sync"
+	"io"
+
+	"github.com/stratosnet/sds/msg/protos"
+)
+
+var (
+	reFileMutex sync.Mutex
+	// key(fileHash) : value(fileSize)
+	rpcFileInfoMap = make(map[string]uint64)
+	// key(fileHash) : value(event)
+	reFileEvent = make(map[string][]string)
+	// key(fileHash) : value(pipe)
+	pipes = make(map[string]pipe)
+)
+
+type pipe struct {
+	reader   *io.PipeReader
+	writer   *io.PipeWriter
+}
+
+// IsFileRpcRemote
+func IsFileRpcRemote(hash string) bool {
+	str := fileMap[hash]
+	if str == "" {
+		return false
+	}
+	return strings.Split(str, ":")[0] == "rpc"
+}
+
+// GetRemoteFileData
+func GetRemoteFileData(hash string, offset *protos.SliceOffset) []byte {
+    // compose event, as well notify the remote user
+	var r []string
+	r = append(r, "1") //return code
+	r = append(r, strconv.FormatUint(offset.SliceOffsetStart, 10)) // extended result: slice start
+	r = append(r, strconv.FormatUint(offset.SliceOffsetEnd, 10)) // extended result: slice end
+
+	// send event and open the pipe for coming data
+    reFileMutex.Lock()
+	reFileEvent[hash] = r
+	var p pipe
+	p.reader, p.writer = io.Pipe()
+	pipes[hash] = p
+    reFileMutex.Unlock()
+
+	// read on pipe
+	data := make([]byte, offset.SliceOffsetEnd - offset.SliceOffsetStart)
+	var cursor []byte
+	var read uint64
+	cursor = data[:]
+	select {
+	case <-time.After(3 * time.Second):
+		return nil
+	default:
+		for {
+			n, err := p.reader.Read(cursor)
+			if err != nil {
+				return nil
+			}
+			read = read + uint64(n)
+			cursor = data[read:]
+			if read >= offset.SliceOffsetEnd - offset.SliceOffsetStart {
+				return []byte(data)
+			}
+		}
+	}
+}
+
+// GetRemoteFileSize
+func GetRemoteFileSize(hash string) uint64{
+	return rpcFileInfoMap[hash]
+}
+
+// SendFileDataBack the rpc handler writes data to slice upload task
+func SendFileDataBack(hash string, content []byte) {
+	reFileMutex.Lock()
+
+	pipes[hash].writer.Write(content)
```

Comment:
direct call value from the map[key] is easy to meet nil pointer, better verify if the value is successfully picked.